upt_macports/upt_macports.py

```diff
@@ -328,3 +330,34 @@ def needs_requirement(self, req, phase):
                  self.standardize_CPAN_version(dep.version) for dep in s])
 
         return super().needs_requirement(req, phase)
+
+    def current_version(self, frontend, pkgname, output=None):
+        '''Return the current version of PKGNAME, packaged using FRONTEND.
+
+        PKGNAME is the name of the package in the upstream package archive, not
+the port name
+        '''
+        self.frontend = frontend.name
+        try:
+            return self.package_versions(pkgname)[0]
+        except:  # noqa
+            self.logger.error(f'Could not get current version for {pkgname}')
+            return super().current_version(frontend, pkgname, output=output)
+
+    def update_package(self, pdiff, output=None):
+        pkg_class = self.pkg_classes[self.frontend]
+        macports_pkg = pkg_class()
+
+        # TODO: This is basically the same code as the one found in
+        # MacPortsPackage._create_output_directories(). It would be nice not to
+        # repeat ourselves.
+        if output is None:
+            pkgname = pdiff.new.name
+            folder_name = macports_pkg._normalized_macports_folder(pkgname)
+            output_dir = os.path.join(macports_pkg.category, folder_name)
+            portfile_path = f'{output_dir}/Portfile'
+        else:
+            portfile_path = output
```

Comment:
I am not sure that I see the use-case for specifying an alternative output here... I'd think that if you want to update a port, you just want to overwrite the Portfile that is currently being used - or? Also, if you specify an output here I would expect it to behave the same as when not using the `-u` flag. In other words if you specify `-o ~/my-macports-clone` it will generate the Portfile in `~/my-macports-clone/python/py-<packagename>/Portfile`. It behaves quite differently here in the sense that you'd need to supply a full path including the `Portfile` part.

Reply:
Hey, thanks for taking a look at the patch!

I'll use PyPI-related examples in my comments, but Rubygems/CPAN work similarly.


> I'd think that if you want to update a port, you just want to overwrite the Portfile that is currently being used - or? 

Yes. The current Portfile is both an input (read the "old" values) and an output (it is overwritten with up-to-date info). upt-macports need to find where the Portfile is, and usually it is really simple: for instance, if a package from PyPI named "requests" must be updated, then the current Portfile must be ./python/py-requests/Portfile. But sometimes, because our naming policy is not bulletproof, this automated "PyPI name to Macports name" does not work, leaving upt-macports unable to find the current Portfile. In that case, you can use "-o /path/to/Portfile" to "help" upt-macports. To summarize, it is usually ok to run:

    $ upt package -f pypi -b macports -u requests

But sometimes, you'll need

    $ upt package -f pypi -b macports -u -o ./python/py-something/Portfile aweirdpackage

Ideally we could find all packages for which the "PyPI name to Macports name" conversion fails and fix them, thus making sure we never have to use -o again.

> In other words if you specify -o ~/my-macports-clone it will generate the Portfile in ~/my-macports-clone/python/py-<packagename>/Portfile. It behaves quite differently here in the sense that you'd need to supply a full path including the Portfile part.

Oh, good point.

My initial idea was that "--output" could be used the same way for creating/updating packages across all upt backends, but somehow it became a bit of a mess. We could make sure that:

    $ upt package -f pypi -b macports -u -o ./python/py-something/Portfile aweirdpackage

and 

    $ upt package -f pypi -b macports -u -o ./python/py-something aweirdpackage

both work, but it would still not be optimal :/